src/datetime.rs

```diff
@@ -246,6 +270,56 @@ impl<Tz: TimeZone> DateTime<Tz> where Tz::Offset: fmt::Display {
         self.format_with_items(ITEMS.iter().cloned()).to_string()
     }
 
+    /// Return an RFC 3339 and ISO 8601 date and time with subseconds
+    /// formatted as per a `SecondsFormat`. If passed `use_z` true and the
+    /// timezone is UTC (offset 0), use 'Z'.
+    pub fn to_rfc3339_opts(&self, subform: SecondsFormat, use_z: bool) -> String {
+        use format::Numeric::*;
+        use format::Pad::Zero;
+        use SecondsFormat::*;
+
+        const PREFIX: &'static [Item<'static>] = &[
+            Item::Numeric(Year, Zero),
+            Item::Literal("-"),
+            Item::Numeric(Month, Zero),
+            Item::Literal("-"),
+            Item::Numeric(Day, Zero),
+            Item::Literal("T"),
+            Item::Numeric(Hour, Zero),
+            Item::Literal(":"),
+            Item::Numeric(Minute, Zero),
+            Item::Literal(":"),
+            Item::Numeric(Second, Zero),
+        ];
+
+        let ssitem = match subform {
+            Secs      => None,
+            Millis    => Some(Item::Fixed(Fixed::Nanosecond3)),
+            Micros    => Some(Item::Fixed(Fixed::Nanosecond6)),
+            Nanos     => Some(Item::Fixed(Fixed::Nanosecond9)),
+            Available => Some(Item::Fixed(Fixed::Nanosecond)),
+        };
+
+        let tzitem = Item::Fixed(
+            if use_z {
+                Fixed::TimezoneOffsetColonZ
+            } else {
+                Fixed::TimezoneOffsetColon
+            }
+        );
+
+        match ssitem {
```

Comment:
this match could probably be something like:

`PREFIX.iter().chain(s.iter()).flat_map(|s| s).chain([tzitem].iter())` I'm not sure off the top of my head what the correct nesting of flat_map with option would be (maybe just sticking the option in the map chain would do it?) but if you're willing to give it a shot I think it'd be cleaner.

Reply:
I tried this, but I couldn't get it to compile.  I'm glad we agree that the format API is hard to use though. :)  I also agree the current impl looks redundant, but in my own attempts to refactor, I ran into various forms of trouble with the barrow-checker.  _Non-Lexical Lifetimes_ might make that possible in the future?